server/src/tree-sitter/analyzer.ts

```diff
@@ -374,73 +375,112 @@ export default class Analyzer {
-        const words = line.split(' ')
-
-        if (new Set(DIRECTIVE_STATEMENT_KEYWORDS).has(words[0])) {
-          logger.debug(`Directive statement keyword found: ${words[0]}`)
-          this.handleKeyword(words[0], line)
-        }
+      // Recursively scan for files in the directive statements `inherit`, `include` and `require`
+      const fileUris = this.getDirectiveFileUris(parsedTree)
+      for (const fileUri of fileUris) {
+        this.sourceIncludeFiles(fileUri)
       }
     } catch (error) {
-      if (error instanceof Error) { // Check if error is an instance of the native JavaScript Error class
-        logger.error(`Error reading file at ${filePath}: ${error.message}`)
+      if (error instanceof Error) {
+        logger.error(`[Analyzer] Error reading file at ${filePath}: ${error.message}`)
       } else if (typeof error === 'string') {
-        logger.error(`Error reading file at ${filePath}: ${error}`)
+        logger.error(`[Analyzer] Error reading file at ${filePath}: ${error}`)
       } else {
-        logger.error(`An unknown error occurred while reading the file at ${filePath}`)
+        logger.error(`[Analyzer] An unknown error occurred while reading the file at ${filePath}`)
       }
     }
   }
 
-  private handleKeyword (keyword: string, line: string): void {
-    const restOfLine: string[] = line.split(keyword).filter(String)
-
-    if (restOfLine.length === 1) {
-      const listOfSymbols: string[] = restOfLine[0].split(' ').filter(String)
-      let definition: Definition = []
-
-      if (listOfSymbols.length === 1) {
-        definition = definition.concat(definitionProvider.createDefinitionForKeyword(keyword, restOfLine[0]))
-      } else if (listOfSymbols.length > 1) {
-        for (const symbol of listOfSymbols) {
-          definition = definition.concat(definitionProvider.createDefinitionForKeyword(keyword, restOfLine[0], symbol))
-        }
-      }
-
-      for (const location of definition) {
-        if (location !== null) {
-          this.extendsFile(this.convertUriStringToFilePath(location.uri))
+  public getDirectiveFileUris (parsedTree: Parser.Tree): string[] {
+    const fileUris: string[] = []
+    parsedTree.rootNode.children.forEach((childNode) => {
+      if (childNode.type === 'inherit_directive') {
+        childNode.children.forEach((n) => {
+          if (n.type === 'inherit_path') {
+            logger.debug(`[Analyzer] Found inherit path: ${n.text}`)
+            const bbclasses = bitBakeProjectScanner.classes.filter((bbclass) => {
+              return bbclass.name === n.text
+            })
+            for (const bbclass of bbclasses) {
+              if (bbclass.path !== undefined) {
+                const uri: string = 'file://' + bbclass.path.dir + '/' + bbclass.path.base
+                fileUris.push(encodeURI(uri))
+              }
+            }
+          }
+        })
+      } else if (childNode.type === 'require_directive' || childNode.type === 'include_directive') {
```

Comment:
Variables definition and classes/includes can come from various other ways like bitbake.conf, .bbappend or BBCLASSEXTEND.
I like the idea of having our own analyzer because getting the real result from bitbake (-e or getVar) is too slow. However we should be mindfull that we are not providing accurate results in some cases.

The other main remark I have is that in Bitbake, a variable has not one unique definition. It can be overriden, appended, modified again in anonymous python, ...
IMO the definition we should point to is the one that is more likely to be next to where the variable is used. Example: extrausers.bbclass defines EXTRA_USERS_PARAMS and also shows how it is used. It won't be the effective value that is displayed, but quickly nabvigating to the relevant source code seems like the most useful purpose of this feature. The other approach would be to point to the latest modification of the variable in order to rather see how the value is defined in the end.

Reply:
`Go to definition` can provide users with a list of "definitions". See this example screenshot:
![image](https://github.com/savoirfairelinux/vscode-bitbake/assets/47988425/24ec7c59-6b92-4fd1-b564-6888679699b0)
A small window with the list of possible definitions appears when users try to find the definitions. (If there is only one, users will be redirected to the source code immediately)

Do you think it makes sense to give users all possible definitions? However, the order of the list can't represent the modification history as it is only known by bitbake.

Reply:
That's great! It may not be exhaustive though because of the other ways to include files but it's a great start that should catch most interesting definitions.
In that case, make sure to also catch operators and overrides definitions like
```
MYVAR:append:thatoverride = "abc"
```
And have them in that list.
If Ctrl+Click picks one of the list automatically, make sure it picks the first.

Reply:
See https://github.com/savoirfairelinux/vscode-bitbake/pull/99/commits/f5cb5bb9c6e8ebda0cb27e565ae36ad75670e221
I can later add other possible sources of the definition that you mentioned above